lib/features/thread/domain/usecases/move_multiple_email_to_mailbox_interactor.dart

```diff
@@ -0,0 +1,42 @@
+import 'package:core/core.dart';
+import 'package:dartz/dartz.dart';
+import 'package:jmap_dart_client/jmap/account_id.dart';
+import 'package:jmap_dart_client/jmap/mail/email/email.dart';
+import 'package:model/email/move_action.dart';
+import 'package:model/mailbox/presentation_mailbox.dart';
+import 'package:tmail_ui_user/features/email/domain/model/move_request.dart';
+import 'package:tmail_ui_user/features/email/domain/usecases/move_to_mailbox_interactor.dart';
+import 'package:tmail_ui_user/features/thread/domain/state/move_multiple_email_to_mailbox_state.dart';
+
+class MoveMultipleEmailToMailboxInteractor {
```

Comment:
class MoveMultipleEmailToMailboxInteractor {
  final MoveToMailboxInteractor _moveToMailboxInteractor;

  MoveMultipleEmailToMailboxInteractor(this._moveToMailboxInteractor);

  Stream<Either<Failure, Success>> execute(
      AccountId accountId,
      List<EmailId> listEmailId,
      PresentationMailbox currentMailbox,
      PresentationMailbox destinationMailbox,
      MoveAction moveAction,
  ) async* {
    try {
      yield Right<Failure, Success>(UIState.loading);
      
      final streamBuilder = StreamController<Either<Failure, Success>>();
      
      listEmailId.map((emailId) => streamBuilder.addStream(
        _moveToMailboxInteractor.execute(
          accountId,
          MoveRequest(emailId, currentMailbox, destinationMailbox, moveAction)))
      );

      final listResult = await streamBuilder.stream
        .where(_validateSuccessState)
        .toList();
      
      streamBuilder.close();
      
      if (listEmailId.length == listResult.length) {
        yield Right(MoveMultipleEmailToMailboxAllSuccess(listEmailId, currentMailbox, destinationMailbox, moveAction));
      } else if (listResult.length == 0) {
        yield Left(MoveMultipleEmailToMailboxAllFailure(moveAction));
      }
      yield Right(MoveMultipleEmailToMailboxHasSomeEmailFailure(listEmailId, currentMailbox, destinationMailbox, moveAction));
    } catch (e) {
      yield Left(MoveMultipleEmailToMailboxFailure(e, moveAction));
    }
  }

  bool _validateSuccessState(Either<Failure, Success> state) {
    if (state.isRight() && state is UIState) {
      if (state is MoveToMailboxSuccess) {
        return true;
      }
    }
    return false;
  }
}

Reply:
Please don't make miscellaneous comments here. Please give a possible solution. Be a worthy leader.

Reply:
:), please pair it tmr. We need to find the better solution for it.

Reply:
sorry :), I come back with better version: need to use `async` package:

```dart
import 'dart:async';

import 'package:async/async.dart';
import 'package:core/core.dart';
import 'package:dartz/dartz.dart';
import 'package:jmap_dart_client/jmap/account_id.dart';
import 'package:jmap_dart_client/jmap/mail/email/email.dart';
import 'package:model/email/move_action.dart';
import 'package:model/mailbox/presentation_mailbox.dart';
import 'package:tmail_ui_user/features/email/domain/model/move_request.dart';
import 'package:tmail_ui_user/features/email/domain/state/move_to_mailbox_state.dart';
import 'package:tmail_ui_user/features/email/domain/usecases/move_to_mailbox_interactor.dart';
import 'package:tmail_ui_user/features/thread/domain/state/move_multiple_email_to_mailbox_state.dart';

class MoveMultipleEmailToMailboxInteractor {
  final MoveToMailboxInteractor _moveToMailboxInteractor;

  MoveMultipleEmailToMailboxInteractor(this._moveToMailboxInteractor);

  Stream<Either<Failure, Success>> execute(
      AccountId accountId,
      List<EmailId> listEmailId,
      PresentationMailbox currentMailbox,
      PresentationMailbox destinationMailbox,
      MoveAction moveAction,
  ) async* {
    try {
      yield Right<Failure, Success>(UIState.loading);
      
      final mergedStream = StreamGroup.merge(
        listEmailId.map((emailId) => _moveToMailboxInteractor.execute(
            accountId,
            MoveRequest(emailId, currentMailbox, destinationMailbox, moveAction))
      ));
      
      final listResult = await mergedStream
        .where(_validateSuccessState)
        .toList();
      
      if (listEmailId.length == listResult.length) {
        yield Right(MoveMultipleEmailToMailboxAllSuccess(listEmailId, currentMailbox, destinationMailbox, moveAction));
      } else if (listResult.length == 0) {
        yield Left(MoveMultipleEmailToMailboxAllFailure(moveAction));
      }
      yield Right(MoveMultipleEmailToMailboxHasSomeEmailFailure(listEmailId, currentMailbox, destinationMailbox, moveAction));
    } catch (e) {
      yield Left(MoveMultipleEmailToMailboxFailure(e, moveAction));
    }
  }

  bool _validateSuccessState(Either<Failure, Success> state) {
    if (state.isRight() && state is UIState) {
      if (state is MoveToMailboxSuccess) {
        return true;
      }
    }
    return false;
  }
}
```